portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -34,51 +40,67 @@
-      Date date = new Date();
-      SimpleDateFormat formatter = new SimpleDateFormat("MMM d, HH:MM a");
+    public Comment(String user, String content, String time) {
       this.user = user;
       this.content = content;
-      this.time = formatter.format(date);
+      this.time = time;
     }
   }
 
-  private List<Comment> text;
-
-  @Override
-  public void init() {
-    text = new ArrayList<>();
-  }
+  private int numCommentsDisplayed = 10;
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String my_json = convertToJson(text);
+    int count = 0;
+    Query myQuery = new Query("Task").addSort("timeStamp", SortDirection.DESCENDING);
+    DatastoreService dataStore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = dataStore.prepare(myQuery);
+    List<Comment> commentList = new ArrayList<>();
+
+    for (Entity entity : results.asIterable()) {
+      String user = (String) entity.getProperty("user");
+      String content = (String) entity.getProperty("content");
+      String displayTime = (String) entity.getProperty("displayTime");
+
+      Comment myComment = new Comment(user, content, displayTime);
+      commentList.add(myComment);
+
+      count++;
+      if (count >= numCommentsDisplayed) {
+        break;
+      }
+    }
+    String myJson = convertToJson(commentList);
     response.setContentType("application/json;");
-    response.getWriter().println(my_json);
+    response.getWriter().println(myJson);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Entity myEntity = new Entity("Task");
-
-    String user = request.getParameter("user");
-    String content = request.getParameter("content");
-    long time_stamp = System.currentTimeMillis();
-
-    myEntity.setProperty("user", user);
-    myEntity.setProperty("content", content);
-    myEntity.setProperty("time_stamp", time_stamp);
-
-    DatastoreService data_store = DatastoreServiceFactory.getDatastoreService();
-    data_store.put(myEntity);
+    numCommentsDisplayed = setMax(request);
+    response.sendRedirect("/chat.html");
+  }
 
-    // redirect to same page so that the page refreshes with new comment
-    // response.sendRedirect("/chat.html");
+  private String convertToJson(List<Comment> commentList) {
+    Gson myGson = new Gson();
+    String myJson = myGson.toJson(commentList);
+    return myJson;
   }
 
-  private String convertToJson(List<Comment> text) {
-    Gson my_gson = new Gson();
-    String my_json = my_gson.toJson(text);
-    return my_json;
+  private int setMax(HttpServletRequest request) {
+    String mynum = request.getParameter("numComments");
+    int num;
+    // safeguards the program just in case the user does not enter a number
+    try {
+      num = Integer.parseInt(mynum);
+    } catch (NumberFormatException e) {
+      return 3;
```

Comment:
There are a bunch of magic numbers in this method; can we turn them into class constants like `numCommentsDisplayed` above?

Reply:
I made the 3 into a constant but got rid of the other numbers because the html form attributes were set to not allow any numbers not within my specified range.